src/arch/riscv/rvdbg013.c

```diff
@@ -0,0 +1,1772 @@
+		DEBUG_INFO("Workaround for single stepping ESP32-C3\n");
+		cause = 4;
+	}
+	DEBUG_TARGET("DCSR 0x%08" PRIx32 ", cause = %d\n", dcsr,cause);
+	if (cause == 0)
+		return TARGET_HALT_RUNNING;
+	switch (cause) {
+	case 1: /* Software breakpoint */
+	case 2: /* Hardware trigger breakpoint */
+		return TARGET_HALT_BREAKPOINT;
+	case 3: return TARGET_HALT_REQUEST;
+	case 4: return TARGET_HALT_STEPPING;
+	case 5: return TARGET_HALT_REQUEST;
+	default:
+		return TARGET_HALT_ERROR;
+	}
+}
+
+static int riscv_breakwatch_set(target *t, struct breakwatch *bw)
+{
+	uint32_t mcontrol = CSR_MCONTROL_DMODE | CSR_MCONTROL_ACTION_DEBUG |
+		CSR_MCONTROL_ENABLE_MASK;
+
+	switch (bw->type) {
+	case TARGET_BREAK_HARD:
+		mcontrol |= CSR_MCONTROL_X;
+		break;
+	case TARGET_WATCH_WRITE:
+		mcontrol |= CSR_MCONTROL_W;
+		break;
+	case TARGET_WATCH_READ:
+		mcontrol |= CSR_MCONTROL_R;
+		break;
+	case TARGET_WATCH_ACCESS:
+		mcontrol |= CSR_MCONTROL_RW;
+		break;
+	default:
+		return 1;
+	}
+
+	uint32_t tselect_saved;
+	rvdbg_reg_read(t, HART_REG_CSR_TSELECT, &tselect_saved, 4);
+
+	uint32_t i;
+	for (i = 0; ; i++) {
+		rvdbg_reg_write(t, HART_REG_CSR_TSELECT, &i, 4);
+		uint32_t tselect;
+		rvdbg_reg_read(t, HART_REG_CSR_TSELECT, &tselect, 4);
+		if (tselect != i)
+			return -1;
+		uint32_t tdata1;
+		rvdbg_reg_read(t, HART_REG_CSR_MCONTROL, &tdata1, 4);
+		uint8_t type = (tdata1 >> (32-4)) & 0xf;
+		if ((type == 0))
+			return -1;
+		if ((type == 2)  &&
+			(((tdata1 & CSR_MCONTROL_RWX) == 0) ||
+			 ((tdata1 & CSR_MCONTROL_ENABLE_MASK) == 0)))
+			break;
+	}
+	/* if we get here tselect = i is the index of our trigger */
+	bw->reserved[0] = i;
+
+	rvdbg_reg_write(t, HART_REG_CSR_MCONTROL, &mcontrol, 4);
+	rvdbg_reg_write(t, HART_REG_CSR_TDATA2, &bw->addr, 4);
+
+	/* Restore saved tselect */
+	rvdbg_reg_write(t, HART_REG_CSR_TSELECT, &tselect_saved, 4);
+	return 0;
+}
+
+static int riscv_breakwatch_clear(target *t, struct breakwatch *bw)
+{
+	uint32_t i = bw->reserved[0];
+	uint32_t tselect_saved;
+	rvdbg_reg_read(t, HART_REG_CSR_TSELECT, &tselect_saved, 4);
+
+	rvdbg_reg_write(t, HART_REG_CSR_TSELECT, &i, 4);
+	i = 0;
+	rvdbg_reg_write(t, HART_REG_CSR_MCONTROL, &i, 4);
+
+	/* Restore saved tselect */
+	rvdbg_reg_write(t, HART_REG_CSR_TSELECT, &tselect_saved, 4);
+	return 0;
+}
+
+int rvdbg_dmi_init(RVDBGv013_DMI_t *dmi)
+{
+	uint8_t version;
+	uint32_t dmstatus, nextdmaddr, dmcontrol;
+	target *t;
+
+    DEBUG_INFO("  debug version = %s\n  abits = %d\n  idle = ",
+		rvdbg_version_tostr(dmi->debug_version), dmi->abits);
+
+	dmi->error = false;
+
+	switch (dmi->idle) {
+		case 0:
+			DEBUG_INFO("no run/test state\n");
+			break;
+		case 1:
+			DEBUG_INFO("leave run/test immediately\n");
+			break;
+		default:
+			DEBUG_INFO("stay %d cycles in run/test\n", dmi->idle - 1);
+			break;
+	}
+
+	dmi->rvdbg_dmi_reset(dmi, false);
+	/* Reset DM to initial values.
+	 * 0,13 to 1.0 incompatible change: Poll dmactive after lowering it. #566
```

Comment:
we already know the spec version here, maybe do the poll optionally depending on the reported version?